woocommerce/woocommerce
# Optimize Workflow Caching

### All Submissions:


### Changes proposed in this Pull Request:

This pull request centralizing the dependency installation, building, and caching within a single composite action. Since virtually every single workflow runs both `cache-deps` and `install-build`, it seems reasonable that these should instead be encapsulated in a single action. This pull request also adds support for caching build outputs with the intention of dramatically speeding up related CI checks.

Closes #33908.

### How to test the changes in this Pull Request:

1. Confirm that all CI checks are passing in this PR.
2. Confirm that the "Setup WooCommerce Monorepo" steps in all workflows pulls from the cache successfully.

Once you've confirmed that things seem to be working here, there's some steps you should take in a fork of `woocommerce/woocommerce`.

1. Confirm that the first run of CI does not populate from cache. The "Setup WooCommerce Monorepo" step should take a while to run.
2. Confirm that the second run of CI populates from the cache. The "Setup WooCommerce Monorepo" step should run quickly.
3. Merge the contents of this PR into your fork.
4. Verify that the "Prime Cache" workflow runs correctly
5. Create a new PR making a change to WooCommerce's PHP files in a commit in your fork. Confirm that a minimal install and rebuild takes place based on the contents of the PR. This should work because the cache priming in `trunk` saves the day.

Once you've confirmed that the workflows function correctly, you should play around with the inputs on the action to ensure that they work in different cases.

1. `install-filters` supports both single and multi-line filters. One filter per line. This is passed to `pnpm install` and helps indicate what should be installed in the monorepo. This is nice for smaller installations.
2. `build` supports true or false and should skip the Turbo build step if it is false.
3. `build-filters` supports both single and multi-line filters. One filter per line. This is passed to Turborepo's build and is nice to limit the building to a single package or set of packages.
4. `php-version` supports changing the version of PHP that the action sets up.

### Other information:




### FOR PR REVIEWER ONLY:
